chainstate/src/detail/chainstateref/mod.rs

```diff
@@ -762,6 +753,55 @@ impl<'a, S: BlockchainStorageRead, V: TransactionVerificationStrategy> Chainstat
             .collect();
         Ok(result)
     }
+
+    /// Given a new block, obtain its previous block's block index.
+    // Note: the only thing that is specific to new blocks here is the error type PrevBlockNotFound
+    // (which is expected to be produced only for new blocks obtained from peers).
+    // FIXME: is the above note correct?
+    pub fn get_previous_block_index_for_new_block(
+        &self,
+        block: &WithId<Block>,
+    ) -> Result<GenBlockIndex, BlockError> {
+        self.get_gen_block_index(&block.prev_block_id())
+            .map_err(BlockError::PropertyQueryError)?
+            .ok_or(BlockError::PrevBlockNotFound)
+    }
+
+    pub fn new_block_index(
+        &self,
+        block: &WithId<Block>,
+        block_status: BlockStatus,
+    ) -> Result<BlockIndex, BlockError> {
```

Comment:
Note: this is basically the old "add_to_block_index" function, but without checking whether the block index already exists.